pages/blog/posts/understanding-lexical-dynamic-scopes.md

```diff
@@ -0,0 +1,418 @@
+references the latter (from `/properties/name/$ref`). Furthermore,
+`https://example.com/person` references a nested schema resource called
+`https://example.com/item` (from `/$defs/person/$ref`) that is part of an
+external schema resource called `https://example.com/people` shown in the
+bottom left. In the right, a directed graph representation of the relationship
+between these schema resources and the dynamic scope.
+
+Like the other examples so far, the evaluation process starts with the top
+level schema. The dynamic scope at that point is the root schema resource, and
+all other schema resources are out of scope:
+
+![The dynamic scope and remote references (1)](/img/posts/2024/understanding-lexical-dynamic-scopes/dynamic-scope-ref-remote-1.webp)
+
+Because the instance defines a `name` property, we enter the `properties`
+applicator into the subschema at `/properties/name`. This subschema introduces
+a new schema resource. Therefore, the dynamic scope now consists of
+`https://example.com` (the root schema resource) followed by
+`https://example.com/name` (the nested schema resource at `/properties/name`):
+
+![The dynamic scope and remote references (2)](/img/posts/2024/understanding-lexical-dynamic-scopes/dynamic-scope-ref-remote-2.webp)
+
+The `https://example.com/name` schema resource references the other nested
+schema resource: `https://example.com/person`. After following this reference,
+the dynamic scope now consists of `https://example.com` (the root schema
+resource), followed by `https://example.com/name` (the nested schema resource
+at `/properties/name`), followed by `https://example.com/person` (the nested
+schema resource at `/$defs/person`):
+
+![The dynamic scope and remote references (3)](/img/posts/2024/understanding-lexical-dynamic-scopes/dynamic-scope-ref-remote-3.webp)
+
+Now comes an interesting case. We are currently evaluating the nested schema
+resource called `https://example.com/person`. This schema resource points to
+the remote schema called `https://example.com/people` (the `people` part of the
+`people#/item` URI reference), but does not land at its root schema resource.
+Instead, it lands at the nested schema resource called
+`https://example.com/item` (the `#/item` part of the `people#/item` URI
+reference).
+
+In this case, we only push the destination schema resource to the dynamic
+scope, *jumping* over the root schema resource of the external schema.  After
+following this final reference, the dynamic scope now consists of
+`https://example.com` (the root schema resource), followed by
+`https://example.com/name` (the nested schema resource at `/properties/name`),
+followed by `https://example.com/person` (the nested schema resource at
+`/$defs/person`), followed by `https://example.com/item` (the nested schema
+resource of the external schema). Note that the root schema resource called
+`https://example.com/people` is not part of the dynamic scope:
```

Comment:
I don't think that's correct. The reference is `people#/items`. Evaluation has to get `https://example.com/people` before it can follow the pointer (`/items`), and only then does it move into `https://example.com/item`. I think `https://example.com/people` is in the dynamic scope because you had to pass through it to get to `https://example.com/item`. At least that's the way my implementation works.

However, if the reference was just `items`, it would go directly to `https://example.com/items` without passing through `https://example.com/people` first. Although I'm not sure you can count on implementations being able directly resolve to an embedded schema in a completely separate schema document, so I'm not sure just changing it to `items` is a great solution.

Reply:
From what I understood from other discussions, most implementation consist of two steps:

- Scanning the given schema to find identifiers, embedded schema resources, anchors, etc
- Evaluating the given schema

I agree that `https://example.com/people` is *scanned*, but I understood that it is not *evaluated*. Therefore, it is not part of the dynamic scope. 

In any case, I think this is a good point worth clarifying more in the post once we align. @gregsdennis @Relequestual @Julian What is your take on this one?

Reply:
> I agree that `https://example.com/people` is _scanned_, but I understood that it is not _evaluated_.

Yes, that's true.  However, it requires that an implementation scans the schema when it's registered, which may not be the case.  If it doesn't then it doesn't know where to find `items` when it's referenced.

I agree with @jdesrosiers in that with the `people#/items` reference, it has to _get_ `people` first, but I don't think it would be part of the dynamic scope since `items` is a new resource.

> In any case, I think this is a good point worth clarifying more in the post once we align.

This is a sticky nuance that I don't think impacts what the post is trying to say, and it's probably best avoided for this post.  I think we'd be better off with a different kind of example.

(That said, I would like to see some test cases for this to see what different implementations do.  I'll work on that separately.)

My argument against this particular example is that it promotes having a pointer reference cross resource boundary, which is something the spec advises against.

Reply:
I've opened an issue in the test suite to have a case added, but I also think the schema I have there may help this example a bit since it removes the pointer-crossing-resource-boundary issue.

https://github.com/json-schema-org/JSON-Schema-Test-Suite/issues/719

Reply:
Makes sense. Let me simplify this example to remove the pointer reference cross resource boundary. We can avoid touching on this edge case

Reply:
Done! Please take a look at the revised example